lib/vrl/compiler/src/expression/assignment.rs

```diff
@@ -421,6 +425,35 @@ where
         Ok(value)
     }
 
+    fn resolve_void(&self, ctx: &mut Context) -> Void {
+        use Variant::*;
+
+        match self {
+            Single { target, expr } => {
+                let value = expr.resolve(ctx)?;
+                target.insert(value, ctx);
+            }
+            Infallible {
+                ok,
+                err,
+                expr,
+                default,
+            } => match expr.resolve(ctx) {
```

Comment:
I'm having a hard time believing the Rust compiler can do that. There's no way for it to know how the value is used at runtime.

I do agree that the soaks currently don't show the improvement I wanted to see, but I also think we lack some representative test-cases for more complex syntax-heavy VRL programs, so before we merge this PR, I'd like to get one or two additional soaks added, one syntax-heavy variant, and one "simple-function" heavy.

If those don't show any improvements, I'll for sure abandon this PR.

Reply:
> If those don't show any improvements, I'll for sure abandon this PR.

I think that's reasonable.

> I'm having a hard time believing the Rust compiler can do that. There's no way for it to know how the value is used at runtime.

It's pretty trivial to know it doesn't use the result of remap. There's even special Rust syntax we use to say we aren't using the result in the `match`: https://github.com/vectordotdev/vector/blob/370a5a6c6deb57422d4612dbb9ab7c63bf766960/src/transforms/remap.rs#L427-L430
I'm not seeing any boxed traits or anything obvious that might hide this from the compiler. I actually think it's pretty likely this is optimized away automatically already.

Reply:
>  I'm not seeing any boxed traits or anything obvious that might hide this from the compiler. It seems possible this is optimized away automatically already.

All expressions in VRL are `Box<dyn Expression>`. Additionally, the match you're showing there is _only_ for the last expression in the program.

You could be totally right, and that would be great. It would probably be worth it to try and get a minimal example going, and inspect the assembly, but that might be tricky, given how much code gets generated, even for the smallest possible example.

Reply:
> but that might be tricky,

Even if it can inline/optimize it today, that doesn't mean it will continue in the future. If we can get a worst case scenario showing something in a soak test, I'll be convinced. It's just a lot of code to maintain so I want to be sure we are getting something worth it out of it

Reply:
> It's just a lot of code to maintain so I want to be sure we are getting something worth it out of it

Totally agree on that point. The soaks didn't show any difference, so either (1) we don't have soaks that cover use-cases we've seen in bug reports (which I want to validate and/or resolve by adding them), or (2) there really isn't much to be gained here, and we can drop this trial balloon.

Reply:
We could consider changing the interface a bit to reduce the code duplication. Instead of having both `resolve` and `resolve_void`, maybe only use `resolve` and add the options to `Context`, so any function that wants to use this optimization can check and have an if statement for cloning. If it doesn't clone, it can return `Value::null`, similar to how `VoidExpr` behaves today. So you add an additional branch, but that's hopefully insignificant compared to removing a BTreeMap clone.

Just a thought. Overall I like this optimization, I just really don't like how much extra code was added to get it, especially if this is implemented for every function

Reply:
> so any function that wants to use this optimization can check and have an if statement for cloning.

Yeah, that's the original approach I took, but backed out of, as it required an extra conditional check at runtime.

I do agree that it might be insignificant enough to offset the extra cost of code duplication with the current approach.